### PR TITLE
Add index generation automation

### PR DESCRIPTION
**Pull request by bnb**
This _should_ theoretically work on every new push, via Travis. It introduces a slightly modified version of the tooling that's in nodejs/security-wg to address the added directory structure around package names within this repository.

The PR also contains a run of the script and the populated set of data as it should currently look.

There is one caveat that I want to be sure to document here:

If a directory name contains `.json` there will be problems, since that's how the script detects if a directory exists or not. There are other approaches that could accomplish the same goal, but this was the lowest barrier to get the job done. Currently this isn't an issue, and I'm not entirely sure if it's even possible for it to _be_ an issue. That said, I wanted to be sure to document this behavior.

### CODE REVIEW

File: tools/create_index/index_generator.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+const createIndex = function (vulnDirectoryPath, writeDirectoryPath, filename) {
+  const files = fs.readdirSync(vulnDirectoryPath)
+  getVulnDirectoryContents(files, vulnDirectoryPath)
+  writeIndex(vuln, writeDirectoryPath, filename)
+}
+
+
+const getVulnDirectoryContents = function (entries, vulnDir) {
+  // if the first entry's value contains `'.json'` we can assume that all entries are files. Otherwise they're directories.
+  const entriesAreFiles = entries[0].includes('.json')
```
Comment (vdeturckheim) [inline tools/create_index/index_generator.js]: It would be safer to use `stats.isDirectory()` here

File: tools/create_index/index_generator.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  
+  if (entriesAreFiles === true) {
```
Comment (vdeturckheim) [inline tools/create_index/index_generator.js]:
```suggestion
  if (entriesAreFiles) {
```

File: tools/create_index/index_generator.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    console.log(`Files detected. Reading their contents and writing them to ${vulnDir} as directed.`)
+    for(entry of entries) {
+      const filename = entry.slice(-0, entry.toString().indexOf('.json'))
+      
+      const data = fs.readFileSync(vulnDir + entry)
+      
+      createVulnObject(filename, data)
+    }
+  } else if (entriesAreFiles === false) {
```
Comment (vdeturckheim) [inline tools/create_index/index_generator.js]:
```suggestion
  } else {
```
I don't see any other cases

**Review by vdeturckheim — CHANGES_REQUESTED** [review]

**Comment by mhdawson** [comment]
Did we agree on generating and publishing indexes? I remember we agreed to land the tool, but don't remember agreement on publishing the index as part of the data exported by the project. If I just missed that it would be great if you pointed me to the issue with that discussion.  It would probably also be good to reference that anyway for posterity.

**Comment by bnb** [comment]
@mhdawson specifically this: https://github.com/nodejs/security-wg/issues/486#issuecomment-467518562

And the end of this meeting is where we discussed this specific point: [link]( https://www.youtube.com/watch?v=9fWH_rFqyW8&t=50%3A00)

The PR step has been completed in nodejs/security-wg. This PR automatically generates it on each push, which is what we were sticking on at a couple points in the meeting ([link](https://www.youtube.com/watch?v=9fWH_rFqyW8&feature=youtu.be&t=38%3A00), [link](https://youtu.be/9fWH_rFqyW8?t=3120)).

If this is an incorrect approach that's fine. I can leave this open as a strawman if that is helpful to usher a further discussion!

**Comment by mhdawson** [comment]
@bnb a the very end I think you agreed to create a proposal for how the index would be generated/where it would be stored so that we could discuss and come to an agreement.

Sorry for slowing things down but I think we need good consensus before introducing external data that people will depend on and may be hard to stop producing later on....

Now a PR is a concrete proposal, of how and what, I guess I just want to make sure we have consensus for committing to produce this going forward or that we have enough documentation/caveats that we make it clear we are not making any such promises.

**Comment by lirantal** [comment]
Related issue: https://github.com/nodejs/security-wg/issues/115

**Comment by bnb** [comment]
Created an issue: https://github.com/nodejs/security-wg/issues/521

File: index/core.json [hunk 4]
```diff
@@ -0,0 +1 @@
...ion / Denial of Service","overview":"An attacker can cause a Denial of Service (DoS) by establishing an HTTP or HTTPS connection in keep-alive mode and by sending headers very slowly thereby keeping the connection and associated resources alive for a long period of time. Attack potential is mitigated by the use of a load balancer or other proxy layer. This vulnerability is an extension of CVE-2018-12121, addressed in November and impacts all active release lines including 6, 8, 10 and 11."},"61":{"cve":["CVE-2019-5739"],"vulnerable":"6","patched":"^6.17.0","publish_date":"2019-02-28","author":"Matteo Collina","reported_by":"Timur Shemsedinov","ref":"https://nodejs.org/en/blog/vulnerability/february-2019-security-releases/","type":"CWE-400: Uncontrolled Resource Consumption / Denial of Service","overview":"Keep-alive HTTP and HTTPS connections can remain open and inactive for up to 2 minutes in Node.js 6.16.0 and earlier. Node.js 8.0.0 introduced a dedicated server.keepAliveTimeout which defaults to 5 seconds. The behavior in Node.js 6.16.0 and earlier is a potential Denial of Service (DoS) attack vector. Node.js 6.17.0 introduces server.keepAliveTimeout and the 5-second default."}}
```
Comment (sam-github) [inline index/core.json]: the diffs on this are going to be pretty unmanageable, and its hard to review. is it possible to format the index so its got a property or CVE per line? and to terminate it with a newline (to get rid of the red marker you can see above)?

File: tools/create_index/index_generator.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
+var fs = require('fs')
+
+// We will fill in this object with our data
```
Comment (sam-github) [inline tools/create_index/index_generator.js]: This comment isn't so useful, I'd either drop it, or add one that describes the structure of what the object will be filled with.

File: tools/write_index.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+const coreVulnerabilitiesPath = './core/' // directory to read from
+const coreIndexFilename = "core" // name of the file that we'll write, minus `.json`
+
+generator(coreVulnerabilitiesPath, writePath, coreIndexFilename)
+
+const ecosystemVulnerabilitiesPath = './ecosystem/' // directory to read from
+const ecosystemIndexFilename = "ecosystem" // name of the file that we'll write, minus `.json`
+
+generator(ecosystemVulnerabilitiesPath, writePath, ecosystemIndexFilename)
```
Comment (sam-github) [inline tools/write_index.js]: missing NL at end of last line

**Review by sam-github — CHANGES_REQUESTED** [review]
I'm not familiar enough with travis, can you exp

**Comment by sam-github** [comment]
I'm not very familiar with travis, can you explain how this will work? It runs when travis runs, but where does the output go? Does it get added to the PR? How is it that the current state of the repo will be guaranteed to always have up-to-date indices?

**Comment by bnb** [comment]
@sam-github this PR uses the same approach that the rest of this repository uses, which is _on-push_ updates. My understanding is that Travis commits on every push push. @vdeturckheim understands the mechanism better than I do, I just followed the framework that he laid out with the rest of this repo – he'd be the right person to ask 👍

**Comment by sam-github** [comment]
@bnb, fair enough. There do seem to be Travis commits in the history: https://github.com/nodejs/security-advisories/commits/master

Other than my comments, and that I will review the index format after its not all on one line :-), the approach looks OK to me.

File: index/core.json (hunk 4 above)
Comment (bnb) [reply index/core.json]: Made this happen with a one-line change 👍

File: tools/create_index/index_generator.js (hunk 5 above)
Comment (bnb) [reply tools/create_index/index_generator.js]: Improved it 👍

File: tools/create_index/index_generator.js (hunk 1 above)
Comment (bnb) [reply tools/create_index/index_generator.js]: Done.

**Comment by bnb** [comment]
Addressed all comments/feedback! ❤️

File: README.md [hunk 7]
```diff
@@ -6,3 +6,6 @@ Security advisories for Node.js and JavaScript ecosytem [WIP]
 Syncs the Vulnerability database from [nodejs/security-wg](https://github.com/nodejs/security-wg) with this repo.
 
 Please refer to Security WG repo for now (https://github.com/nodejs/security-wg)
+
+#### tools/write_index.js
```
Comment (EiMay) [inline README.md]: https://www.facebook.com/profile.php?id=100028692098727
my fmaily facebook accout old password help me ?

**Comment by vdeturckheim** [comment]
@EiMay I don't understand how/whay you ended up asking this here (I even suspect you are a bot). Please stop polluting this PR or I will ban you from accessing this repo.

**Comment by Trott** [comment]
I've unarchived this repo so I can close all PRs and issues before re-archiving.

**Pull request closed without merging**